pkgs/by-name/vd/vdhcoapp/package.nix

```diff
@@ -0,0 +1,71 @@
+{ lib
+, fetchFromGitHub
+, buildNpmPackage
+, toml2json
+, nodejs
+, ffmpeg
+, callPackage
+, filepicker ? callPackage ./filepicker.nix {}
+, substituteAll
+, makeWrapper
+}:
+
+# This is an adaptation with buildNpmPackage based on https://github.com/milahu/nur-packages/commit/3022ffb3619182ffcd579194e1202e3978e4d55b
+buildNpmPackage rec {
+  pname = "vdhcoapp";
+  version = "2.0.19";
+  srchash = "sha256-8xeZvqpRq71aShVogiwlVD3gQoPGseNOmz5E3KbsZxU=";
+
+  src = fetchFromGitHub {
+    owner = "aclap-dev";
+    repo = "vdhcoapp";
+    rev = "v${version}";
+    hash = srchash;
+  };
+
+  sourceRoot = "${src.name}/app";
+  npmDepsHash = "sha256-E032U2XZdyTER6ROkBosOTn7bweDXHl8voC3BQEz8Wg=";
+  dontNpmBuild = true;
+
+  nativeBuildInputs = [
+    toml2json
+    makeWrapper
+  ];
+
+  patches = [
+    (substituteAll {
+      src = ./ffmpeg-filepicker.patch;
+      inherit ffmpeg;
+      filepicker = lib.getExe filepicker;
+    })
+  ];
+
+  postPatch = ''
+    # Cannot use patch, setting placeholder here
+    substituteInPlace src/native-autoinstall.js \
+      --replace process.execPath "\"${placeholder "out"}/bin/vdhcoapp\""
+  '';
+
+  preBuild = ''
+    toml2json --pretty ../config.toml > src/config.json
+  '';
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/opt/vdhcoapp
+
+    cp -r . "$out/opt/vdhcoapp"
+
+    makeWrapper ${nodejs}/bin/node $out/bin/vdhcoapp \
+      --add-flags $out/opt/vdhcoapp/src/main.js
+    runHook postInstall
+  '';
```

Comment:
How to we install the native hosts files, so that we can use it in combination with the browser? Afaik you will have to install them manually atm, which does not feel very user friendly for nixos users

Reply:
true, this is now done in the postInstall hook. fixed.

Reply:
Did you try to verify the result?
I am quite sure that running the vdhco binaries installer does not do what we expect from it in the nixpkgs context.

I see that the following is happening

```
vdhcoapp> Writing /build/.mozilla/native-messaging-hosts/net.downloadhelper.coapp.json
vdhcoapp> Writing /build/.var/app/org.mozilla.firefox/.mozilla/native-messaging-hosts/net.downloadhelper.coapp.json
```

So it seems that its installing to `$HOME`, which is not valid in our context here as we are  running from a building user that has its `$HOME` set to `/build`.
We could install them system wide, but they are sadly also not valid in out context, so we might want to patch the file to write to `$out`. 
https://github.com/aclap-dev/vdhcoapp/blob/0b40d3e101c429fdae4e02ba1f593c2416b51eff/config.toml#L187 

In addition it seems to have some installation detection (so for which browser it should install the manifests), and i am not sure if that logic is valid in our context here. I would expect the derivation to have all *common* manifests avaiable (so chrome, chromium, firefox, ...), it we need to provide some way that its detecting in which browser context it is installed to generate the right manifest files.

Reply:
> Instead of patching we could also let it write to `$HOME` and move the manifest afterwards to some place, where it would be picked up by the `nativeMessagingHosts` option(s) (hint: idk how that is done for chrome/chromium, but there is an [option](https://search.nixos.org/options?channel=unstable&show=programs.firefox.nativeMessagingHosts.packages&from=0&size=50&sort=relevance&type=packages&query=programs.firefox.nativeMessagingHosts.packages) for at least firefox in nixpkgs and there is also [one in home-manager](https://home-manager-options.extranix.com/?query=programs.firefox.nativeMessagingHosts&release=master), and it would be cool to support those for a good user experience) .

might be kind of a newbee question, but is there a way i could write to those options via my derivation or is the user supposed to do something in the lines of 
```nix
programs.firefox.nativeMassagingHosts = [ "${pkgs.vdhcoapp}/host"]
```
(provided i/this derivation moved the manifest there)

Reply:
> so we might want to patch the file to write to `$out`

this would still not move them to the correct $HOME, right as $out is /nix/store/bla ?

is there a nice way to access the correct user home (e.g. given as input / mkOption), that would be my ansatz

Reply:
you can not write to $HOME of *your* user when building a package derivation, which comes from a build server
I mean we are adding it to nixpkgs, so that it gets shipped by nixos binary cache etc arent we? :slightly_smiling_face: 
But luckly there is a nixos option, which will pick up system wide installed manifests.

`programs.firefox.nativeMessagingHosts.packages`  will look for messaging hosts put at `$out/lib/mozilla/native-messaging-hosts/` of the given packages, via the firefox [wrapper](https://github.com/NixOS/nixpkgs/blob/4c147d4246aa82bd105bd805a6209e4a5071dd32/pkgs/applications/networking/browsers/firefox/wrapper.nix#L38).

Reply:
I have a version that builds the needed manifest files adhoc by itself from the config.toml (or the converted json) adhoc on bulid here https://github.com/Shawn8901/nix-configuration/blob/main/packages/vdhcoapp/default.nix

The problem with your approach is that this results in a strange behavior for hydra builds as the auto-install.js code does look up what browsers are actually installed on the system, thus hydra will likely never place manifests in the store.

I tested the firefox variant with that.

No saying that the code is nice, but possibly it could be a starting point for you

Reply:
Thats a very nice starting point, thanks!

there is no 'nix-way' to add manifest to user installed browsers tho, right?

i think skipping the auto-install and going with your version will be the best way then.

Reply:
I already referenced nix ways of adding using the manifest when provided by the package in my prior comment https://github.com/NixOS/nixpkgs/pull/300051#discussion_r1545344144 

> Instead of patching we could also let it write to $HOME and move the manifest afterwards to some place, where it would be picked up by the nativeMessagingHosts option(s) (hint: idk how that is done for chrome/chromium, but there is an [option](https://search.nixos.org/options?channel=unstable&show=programs.firefox.nativeMessagingHosts.packages&from=0&size=50&sort=relevance&type=packages&query=programs.firefox.nativeMessagingHosts.packages) for at least firefox in nixpkgs and there is also [one in home-manager](https://home-manager-options.extranix.com/?query=programs.firefox.nativeMessagingHosts&release=master), and it would be cool to support those for a good user experience) .

which contains links to the firefox versions.
For chrome/ium i am not aware of builtins for generically adding packages, tho there is a starting point in adding them hardcoded https://github.com/NixOS/nixpkgs/blob/efb96bc312fdd1fc4fdca67b4901f1a3409b7bda/nixos/modules/programs/chromium.nix#L127